File: Solution/126B - Password/126B.cpp
#include <iostream>
#include <vector>
#include <string>
#include <algorithm>
using namespace std;
int main() {
    ios_base::sync_with_stdio(false);
    cin.tie(NULL);

    string s;
    if (!(cin >> s)) return 0;
    int n = s.length();
    vector<int> pi(n, 0);
    int max_mid = 0;

    for (int i = 1; i < n; i++) {
        int j = pi[i - 1];
        while (j > 0 && s[i] != s[j]) j = pi[j - 1];
        if (s[i] == s[j]) j++;
        pi[i] = j;
        
        if (i < n - 1) {
            max_mid = max(max_mid, pi[i]);
        }
    }

    int res = pi[n - 1];

    while (res > max_mid) {
        res = pi[res - 1];
    }

    if (res <= 0) {
        cout << "Just a legend" << "\n";
    } else {
        cout << s.substr(0, res) << "\n";
    }

    return 0;
}

File: Solution/126B - Password/126B.py
import sys

def main():
    s = sys.stdin.readline().strip()
    n = len(s)
    
    pi = [0] * n
    max_mid = 0

    for i in range(1, n):
        j = pi[i - 1]
        while j > 0 and s[i] != s[j]:
            j = pi[j - 1]
        if s[i] == s[j]:
            j += 1
        pi[i] = j

        if i < n - 1:
            max_mid = max(max_mid, pi[i])

    res = pi[n - 1]

    while res > max_mid:
        res = pi[res - 1]

    if res <= 0:
        print("Just a legend")
    else:
        print(s[:res])


if __name__ == "__main__":
    main()